### DISCUSSION TITLE
Run function in background on ESP32?

### DISCUSSION THREAD

**Discussion by dumbnerd08** [micropython/micropython#13212] [ESP32]
I have the following code:
```py
def main():
    doStuff()
    time.sleep(10)
    doOtherStuff()

while True:
    if(stuff):
        main()
        doMoreStuff()
```
I need main to start running and doMoreStuff to run while main is running. How can I do this?

(ESP32 WROOM or smth, v1.22.0-preview.236.ge1a7aa23f on 2023-12-10)

Thanks for your time!

**Comment by davefes** [comment]
https://docs.micropython.org/en/v1.14/library/uasyncio.html
https://github.com/peterhinch/micropython-async

**Comment by sosi-deadeye** [comment]
You should use the Documentation, which fits to the version you use. E.g. all developer versions, are the latest version: https://docs.micropython.org/en/latest/

You can run asyncio together with threads, but not out of the box. But there is a solution: https://github.com/peterhinch/micropython-async/blob/master/v3/docs/THREADING.md#42-more-general-solution

**Comment by shariltumin** [comment]
You can have a look at my [workers framework](https://github.com/shariltumin/workers-framework-micropython). Most people here might see it as anti-pattern. But it might be useful for you, give it a try.
